test/common/upstream/bounded_load_hlb_test.cc

```diff
@@ -0,0 +1,254 @@
+#include <memory>
+
+#include "envoy/config/cluster/v3/cluster.pb.h"
+
+#include "common/upstream/thread_aware_lb_impl.h"
+
+#include "test/common/upstream/utility.h"
+#include "test/mocks/upstream/mocks.h"
+
+namespace Envoy {
+namespace Upstream {
+namespace {
+
+class TestHashingLoadBalancer : public ThreadAwareLoadBalancerBase::HashingLoadBalancer {
+public:
+  TestHashingLoadBalancer(NormalizedHostWeightVector ring) : ring_(std::move(ring)) {}
+  TestHashingLoadBalancer() = default;
+  HostConstSharedPtr chooseHost(uint64_t hash, uint32_t /* attempt */) const override {
+    if (ring_.empty()) {
+      return nullptr;
+    }
+    return ring_.at(hash).first;
+  }
+
+private:
+  const NormalizedHostWeightVector ring_{};
+};
+
+using HostOverloadFactorPredicate = std::function<double(const Host& host, double weight)>;
+class TestBoundedLoadHashingLoadBalancer
+    : public ThreadAwareLoadBalancerBase::BoundedLoadHashingLoadBalancer {
+public:
+  TestBoundedLoadHashingLoadBalancer(
+      ThreadAwareLoadBalancerBase::HashingLoadBalancerSharedPtr hlb_ptr,
+      const NormalizedHostWeightVector& normalized_host_weights, uint32_t hash_balance_factor,
+      HostOverloadFactorPredicate host_overload_factor)
+      : ThreadAwareLoadBalancerBase::BoundedLoadHashingLoadBalancer(
+            hlb_ptr, normalized_host_weights, hash_balance_factor),
+        host_overload_factor_(host_overload_factor) {}
+
+private:
+  HostOverloadFactorPredicate host_overload_factor_;
+  double hostOverloadFactor(const Host& host, double weight) const override {
+    return host_overload_factor_(host, weight);
+  }
+};
+
+class BoundedLoadHashingLoadBalancerTest : public testing::Test {
+public:
+  HostOverloadFactorPredicate
+  getHostOverloadFactorPredicate(const std::vector<std::string>& addresses) {
+    return [addresses](const Host& h, double) -> double {
+      // Each host in the vector gets an increasing overload factor.
+      int index = 1;
+      for (const std::string& host : addresses) {
+        index = index + 1;
+        if (host == h.address()->asString()) {
+          return (100.0 + index) / 100;
+        }
+      }
+      return 0.5;
+    };
+  }
+
+  void createHosts(uint32_t num_hosts, NormalizedHostWeightVector& normalized_host_weights) {
+    const double equal_weight = static_cast<double>(1.0 / num_hosts);
+    for (uint32_t i = 0; i < num_hosts; i++) {
+      normalized_host_weights.push_back(
+          {makeTestHost(info_, fmt::format("tcp://127.0.0.1{}:90", i)), equal_weight});
+    }
+  }
+
+  void createHostsMappedByMultipleHosts(uint32_t num_hosts, NormalizedHostWeightVector& hosts,
```

Comment:
i would give this a clearer name or add a comment, as a reader i dont really get what this is trying to do without reading it all